### ISSUE TITLE
file access with unresolved variables

### ISSUE THREAD

**Issue by az0** [bleachbit/bleachbit#616]
Process Monitor shows PATH NOT FOUND and NAME NOT FOUND errors for many mangled paths. Attached is a list.

Example: `x:\BleachBit-Portable\%$ProgramFiles$$%\Vuze\backup`

Git revision c7a42cafc340bdb1bf517cb7dc74447682980704
Windows 10
[mangled paths.log](https://github.com/bleachbit/bleachbit/files/3548970/mangled.paths.log)

**Comment by mkhon** [comment]
How to reproduce this?

**Comment by az0** [comment]
@mkhon 
I used [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) to monitor BleachBit while it was running. I haven't tried it since August, and much has changed since. I could try again.

**Comment by az0** [comment]
1. Start Procmon64.exe
2. Start BleachBit
3. In Process Monitor, type CTRL+L to edit filters
4. Add filter for _process name_ contains `bleach`
5. Add filter for _path_ contains `$`
6. Click OK

**Comment by mkhon** [comment]
I do not see unexpanded multi-user $$foo$$ variables (like %$ProgramFiles$$%) in process monitor (latest 4.0.1 build).

The only missing expanded variable is $XDG_CACHE_HOME and I think it is ok (e.g. from google-chome.xml).

**Comment by az0** [comment]
Yes, I no longer see the multi-value variables with `$$`, but I do see these on startup

```
C:\Users\username\Downloads\BleachBit-4.0.1.1650-portable\BleachBit-Portable\$XDG_CACHE_HOME\chromium
C:\Users\username\Downloads\BleachBit-4.0.1.1650-portable\BleachBit-Portable\$profile$\bookmarkbackups\
C:\Users\username\Downloads\BleachBit-4.0.1.1650-portable\BleachBit-Portable\$localapp$\Cache
C:\Users\username\Downloads\BleachBit-4.0.1.1650-portable\BleachBit-Portable\$ProgramFiles$\Vuze\backup
```

As implied by the paths, this is build 1650 in portable mode on Windows 10.

Besides wasting some time, this doesn't seem to break anything.

**Comment by az0** [comment]
On Ubuntu with revision 7b6c29b44706d4c003df11619c74fa805d0df7ce , I get something similar 

```bash
strace python3 bleachbit.py &>/tmp/strace.log
grep ENOENT /tmp/strace.log   | grep "\\$"
```

With similar lines removed, here are some entries
```
lstat("$USERPROFILE\\Application Data\\Google\\GoogleEarth\\dbCache.dat", 0x7ffe95df5bd0) = -1 ENOENT (No such file or directory)
lstat("$APPDATA\\Google\\Local Search History\\", 0x7ffe95df5a70) = -1 ENOENT (No such file or directory)
lstat("$localappdata\\Mozilla\\Profiles\\default\\Cache\\", 0x7ffe95df5a70) = -1 ENOENT (No such file or directory)
lstat("$$ProgramFiles$$\\backup\\", 0x7ffe95df5a70) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "$$profile$$/bookmarkbackups", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = -1 ENOENT (No such file or directory)
lstat("$$profile$$/netpredictions.sqlite", 0x7ffe95df5bd0) = -1 ENOENT (No such file or directory)
```

**Comment by mkhon** [comment]
@az0 I believe these references to Windows variables (both regular vars like $APPDATA and mult-value like $$ProgramFiles$$) when running on Linux are ok

**Comment by az0** [comment]
Quickly-made benchmark on Linux with 90 cleaners with code as of 32dd1d4968d9062b83186a6e1bbe70e995e375fb  :  filtering unresolved variables reduced half of file system calls. Each call is cheap, but there were >10K calls. On fast SSD with warm cache, there is negligible benefit. **With Linux, SSD, and cold cache, filtering saved ~50ms.**

I didn't test spinning HDD, Windows, or Winapp2.ini (which adds more cleaners). **The HDD and Winapp2.ini would likely show greater benefit to filtering the unresolved variables.**
